### PR TITLE
refactor(marketprice): Make invalid state unrepresentable

### PR DESCRIPTION
**Pull request by KirilMihaylov**

### CODE REVIEW

File: packages/marketprice/src/alarms/mod.rs [hunk 1]
```diff
@@ -181,28 +181,54 @@ impl<'storage, S> PriceAlarms<'storage, S>
 where
     S: Deref<Target = dyn Storage + 'storage> + DerefMut,
 {
+    pub fn add_alarms<C, BaseC>(
```
Comment (gmanev7) [inline packages/marketprice/src/alarms/mod.rs:184]: in singular please

File: packages/marketprice/src/alarms/mod.rs [hunk 2]
```diff
@@ -181,28 +181,54 @@ impl<'storage, S> PriceAlarms<'storage, S>
...
+        &mut self,
+        subscriber: Addr,
+        below_alarm: Price<C, BaseC>,
+        above_or_equal_alarm: Option<Price<C, BaseC>>,
```
Comment (gmanev7) [inline packages/marketprice/src/alarms/mod.rs:188]: remove the parasite `_alarm`

File: packages/marketprice/src/alarms/mod.rs [hunk 3]
```diff
@@ -181,28 +181,54 @@ impl<'storage, S> PriceAlarms<'storage, S>
...
         BaseC: Currency,
     {
-        self.add_alarm_above_or_equal_internal(subscriber, &NormalizedPrice::new(&alarm))
+        self.add_alarm_below_internal(subscriber.clone(), &NormalizedPrice::new(&below_alarm))
+            .and_then(|()| {
+                self.add_alarm_above_or_equal_internal(
+                    subscriber,
+                    &NormalizedPrice::new(&above_or_equal_alarm),
+                )
+            })
```
Comment (gmanev7) [inline packages/marketprice/src/alarms/mod.rs:231]: just merge the two cases into a common implementation modulo an Option value

File: packages/marketprice/src/alarms/mod.rs [hunk 4]
```diff
@@ -262,29 +258,12 @@ where
...
-                })
-            })
+        self.pop_front_in_delivery("Received success reply status")
+            .map(|_: AlarmWithSubscriber| ())
     }
 
     pub fn last_failed(&mut self) -> Result<(), AlarmError> {
-        self.in_delivery
-            .pop_front(self.storage.deref_mut())
-            .map_err(AlarmError::LastFailedRemove)
```
Comment (gmanev7) [inline packages/marketprice/src/alarms/mod.rs:280]: I am afraid we misled the caller by sending back Error::LastDeliveredRemove instead

Comment (KirilMihaylov) [reply packages/marketprice/src/alarms/mod.rs:280]: Good catch!

**Review by gmanev7 — APPROVED** [review]

**Pull request merged**